File: MAP(Advanced Programming Methods)/LabAssig7/LabAssig7/Model/Statements/ReadFileStatement.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using LabAssig7.Model.ADT;
using LabAssig7.Model.Expressions;
using LabAssig7.Model;
using System.IO;
namespace LabAssig7.Model.Statements
{
    public class ReadFileStatement:Statement
    {
        private Expression expr;
        private string varName;

        public ReadFileStatement(Expression e,string var)
        {
            this.expr = e;
            this.varName = var;
        }

        public PrgState Execute(PrgState p)
        {
            int res = this.expr.Eval(p.Dict);
            if (!p.FileTable.Contains(res))
                throw new SystemException("File not found");
            FileData fd = p.FileTable.GetValue(res);
            StreamReader head = fd.Header;
            string line = head.ReadLine();
            Expression val;
            if (line == null)
                val = new ConstantExpression(0);
            else
                val = new ConstantExpression(Int32.Parse(line));
            Statement st = new AssignStatement(this.varName, val);
            st.Execute(p);
            return p;

        }
        public override string ToString()
        {
            return "ReadFileStatement(" + this.varName + "," + this.expr + ")";
        }
    }
}

File: MAP(Advanced Programming Methods)/lab3/src/Model/Statement/ReadFileStatement.java
package Model.Statement;

import Exceptions.FileException;
import Model.ADT.IFileTable;
import Model.Expression.ConstantExpression;
import Model.Expression.Expression;
import Model.FileData;
import Model.PrgState;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;

public class ReadFileStatement implements Statement {

    private Expression fileDescriptor;
    private String varName;

    public ReadFileStatement(Expression fd,String var)
    {
        this.fileDescriptor=fd;
        this.varName=var;
    }

    @Override
    public PrgState execute(PrgState p)
    {
        try
        {
            int descriptor=this.fileDescriptor.eval(p.getSymbolTable());
            IFileTable<Integer, FileData> filetable=p.getFileTable();
            if(!filetable.contains(descriptor))
                throw new FileException("Descriptor does not match any open file");
            BufferedReader buff= filetable.get(descriptor).getHeader();
            String line = buff.readLine();
            ConstantExpression val;
            if(line==null)
                val = new ConstantExpression(0);
            else
                val=new ConstantExpression(Integer.parseInt(line));

            Statement st = new AssignStatement(this.varName,val);
            st.execute(p);
            return p;

        }
        catch (IOException e)
        {
            throw  new FileException(e.toString()+"Eroare read file");
        }


    }
}
